src/AdapterInterface.php

```diff
@@ -184,6 +184,43 @@ public function write($fileDescriptor, $data, $length, $offset);
      */
     public function close($fd);
 
+    /**
+     * Reads the entire file.
+     *
+     * This is an optimization for adapters which can optimize
+     * the open -> write -> close sequence into one call.
+     *
+     * @param string $path
+     * @param int $offset
+     * @param int|null $length
+     * @return PromiseInterface
+     */
+    public function getContents($path, $offset = 0, $length = null);
+
+    /**
+     * Writes directly to the file.
```

Comment:
LGTM! Maybe add some distinction and/or link wrt `appendContents()`?

"Will overwrite the file with the given contents if it does exist. Will create the file with the given contents if does not exist." ?